cognite_toolkit/_cdf_tk/utils/cdf.py

```diff
@@ -115,28 +115,51 @@ def iterate_instances(
         body["cursor"] = next_cursor
 
 
+@overload
 def read_auth(
+    authentication: object,
+    client_config: ToolkitClientConfig,
+    identifier: Hashable,
+    resource_name: str,
+    allow_oidc: Literal[False] = False,
+    console: Console | None = None,
+) -> ClientCredentials: ...
+
+
+@overload
+def read_auth(
+    authentication: object,
+    client_config: ToolkitClientConfig,
+    identifier: Hashable,
+    resource_name: str,
+    allow_oidc: Literal[True],
+    console: Console | None = None,
+) -> OidcCredentials: ...
+
+
+def read_auth(
+    authentication: object,
+    client_config: ToolkitClientConfig,
     identifier: Hashable,
-    resource: dict[str, Any],
-    client: ToolkitClient,
     resource_name: str,
+    allow_oidc: bool = False,
     console: Console | None = None,
-) -> ClientCredentials:
-    auth = resource.get("authentication")
-    if auth is None:
-        if client.config.is_strict_validation or not isinstance(client.config.credentials, OAuthClientCredentials):
+) -> ClientCredentials | OidcCredentials:
+    if authentication is None:
+        if client_config.is_strict_validation or not isinstance(client_config.credentials, OAuthClientCredentials):
             raise ToolkitRequiredValueError(f"Authentication is missing for {resource_name} {identifier!r}.")
         else:
             HighSeverityWarning(
                 f"Authentication is missing for {resource_name} {identifier!r}. Falling back to the Toolkit credentials"
             ).print_warning(console=console)
-        credentials = ClientCredentials(client.config.credentials.client_id, client.config.credentials.client_secret)
-    elif not isinstance(auth, dict):
+        return ClientCredentials(client_config.credentials.client_id, client_config.credentials.client_secret)
+    elif not isinstance(authentication, dict):
         raise ToolkitTypeError(f"Authentication must be a dictionary for {resource_name} {identifier!r}")
-    elif "clientId" not in auth or "clientSecret" not in auth:
+    elif "clientId" not in authentication or "clientSecret" not in authentication:
         raise ToolkitRequiredValueError(
             f"Authentication must contain clientId and clientSecret for {resource_name} {identifier!r}"
         )
+    elif allow_oidc and "tokenUri" in authentication and "cdfProjectName" in authentication:
```

Comment:
this function misbehaves with types in edge cases, contract is that return type depends on allow_oids, but in several cases ClientCredentials can be returned with allow_oidc=true

Reply:
That is a bug in the overload. When allow_oidc=true it is still ok to have it return `ClientCredentials`. Fix it.